experimental/JuLie/partitions.jl

```diff
@@ -401,19 +403,95 @@ function ascending_partitions(n::IntegerUnion; alg="ks")
 
+################################################################################
+# Generating and counting restricted partitions
+################################################################################
+
+@doc Markdown.doc"""
+    num_partitions(n::IntegerUnion, k::IntegerUnion)
+
+The number of integer partitions of the non-negative integer `n` into `k >= 0` parts. 
+
+# Algorithm
+We use the recurrence relation ``p_k(n) = p_{k-1}(n-1) + p_k(n-k)``, where ``p_k(n)``
+denotes the number of partitions of ``n`` into ``k`` parts; see [Knu11](@cite), Section
+7.2.1.4, Equation (39) on page 399.
+
+# References
+1. [OEIS](@cite), [A008284](https://oeis.org/A008284)
+"""
+function num_partitions(n::IntegerUnion, k::IntegerUnion)
+  
+  @req n >= 0 "n >= 0 required"
+  @req k >= 0 "k >= 0 required"
+  
+  n = ZZ(n)
+  k = ZZ(k)
+
+  # Special cases
+  if n == k
+    return ZZ(1)
+  elseif n < k || k == 0
+    return ZZ(0)
+  elseif k == 1
+    return ZZ(1)
+
+  # See https://oeis.org/A008284
+  elseif n < 2*k
+    return num_partitions(n-k) #n-k>=0 holds since the case n<k was already handled
+
+  # See https://oeis.org/A008284
+  elseif n <= 2+3*k
+    p = num_partitions(n-k) #n-k>=0 holds since the case n<k was already handled
+    for i=0:Int(n)-2*Int(k)-1
+      p = p - num_partitions(ZZ(i))
+    end
+    return p
+
+  # Otherwise, use recurrence.
+  # The following is taken from the GAP code in lib/combinat.gi
+  # It uses the standard recurrence relation but in a more intelligent
+  # way without recursion.
+  else
+    n = Int(n)
+    k = Int(k)
+    p = fill( ZZ(1), n )
+    for l = 2:k
+      for m = l+1:n-l+1
+        p[m] = p[m] + p[m-l]
+      end
+    end
+    return p[n-k+1]
+  end
 
+end
 
 @doc Markdown.doc"""
-    partitions(m::IntegerUnion, n::IntegerUnion, l1::IntegerUnion, l2::IntegerUnion; only_distinct_parts::Bool = false)
+    partitions(m::T, n::IntegerUnion, l1::IntegerUnion, l2::IntegerUnion; only_distinct_parts::Bool = false) where T <: IntegerUnion
 
-A list of all partitions of an integer ``m ≥ 0`` into ``n ≥ 0`` parts with
-lower bound ``l1 ≥ 0`` and upper bound ``l2 ≥ l1`` for the parts. There are
-two choices for the parameter `only_distinct_parts`:
+A list of all partitions of a non-negative integer `m` into `n >= 0` parts with lower bound
+`l1 >= 0` and upper bound `l2` for the parts. There are two choices for the parameter
+`only_distinct_parts`:
 * `false`: no further restriction (*default*);
-* `true`: only distinct parts.
-The partitions are produced in *decreasing* order.
+* `true`: only distinct parts. The partitions are produced in *decreasing* order.
 
-The algorithm used is "parta" in [RJ76](@cite), de-gotoed from old ALGOL code by E. Thiel!
+# Examples
+```jldoctest
+# All partitions of 7 into 3 parts where all parts are between 1 and 4.
+julia> partitions(7, 3, 1, 4)
+3-element Vector{Partition{Int64}}:
+ [4, 2, 1]
+ [3, 3, 1]
+ [3, 2, 2]
+
+# Same as above but requiring all parts to be distinct.
```

Comment:
You cannot put comments like this: what you are effectively expressing here is that the output of the preceding command (`partitions(7, 3, 1, 4)`) ends with the text `Same as a ...` which it doesn't, hence the doctests fail.

So just move the comment out of the doctest and make it part of the regular text: 
````suggestion
```
Same as above but requiring all parts to be distinct:
```jldoctest
````